gcc8 emits -Wcast-function-type when compiling any module with a function binding.

## Issue description

gcc8 emits -Wcast-function-type (enabled by -Wextra) when compiling any module with a function binding.

## Reproducible example code

Start with pybind11/python_example, replace `src/main.cpp`'s contents with
```
#include <pybind11/pybind11.h>

namespace py = pybind11;

PYBIND11_MODULE(python_example, m) {
    m.def("foo", []() -> void {});
}
```
and compile with gcc8 and CFLAGS=-Wextra.  One gets
```
running build
running build_ext
gcc -pthread -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -Wextra -fPIC -I/usr/include/python3.6m -c /tmp/tmpk61gqk0m.cpp -o tmp/tmpk61gqk0m.o -std=c++14
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
gcc -pthread -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -Wextra -fPIC -I/usr/include/python3.6m -c /tmp/tmpre7j64d4.cpp -o tmp/tmpre7j64d4.o -fvisibility=hidden
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
building 'python_example' extension
gcc -pthread -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -Wextra -fPIC -I/usr/include/python3.6m -I/home/antony/.local/include/python3.6m -I/usr/include/python3.6m -c src/main.cpp -o build/temp.linux-x86_64-3.6/src/main.o -DVERSION_INFO="0.0.1" -std=c++14 -fvisibility=hidden
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
In file included from src/main.cpp:1:
/usr/include/python3.6m/pybind11/pybind11.h: In member function ‘void pybind11::cpp_function::initialize_generic(pybind11::detail::function_record*, const char*, const std::type_info* const*, pybind11::size_t)’:
/usr/include/python3.6m/pybind11/pybind11.h:317:74: warning: cast between incompatible function types from ‘PyObject* (*)(PyObject*, PyObject*, PyObject*)’ {aka ‘_object* (*)(_object*, _object*, _object*)’} to ‘PyCFunction’ {aka ‘_object* (*)(_object*, _object*)’} [-Wcast-function-type]
             rec->def->ml_meth = reinterpret_cast<PyCFunction>(*dispatcher);
                                                                          ^
g++ -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wextra build/temp.linux-x86_64-3.6/src/main.o -L/usr/lib -lpython3.6m -o build/lib.linux-x86_64-3.6/python_example.cpython-36m-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-3.6/python_example.cpython-36m-x86_64-linux-gnu.so ->
```

It looks like CPython is solving this by casting to a `void*` before casting to PyCFunction (https://bugs.python.org/issue33012 and PRs linked therein).

--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -306,7 +306,7 @@ protected:
             rec->def = new PyMethodDef();
             std::memset(rec->def, 0, sizeof(PyMethodDef));
             rec->def->ml_name = rec->name;
-            rec->def->ml_meth = reinterpret_cast<PyCFunction>(*dispatcher);
+            rec->def->ml_meth = reinterpret_cast<PyCFunction>(reinterpret_cast<void*>(*dispatcher));
             rec->def->ml_flags = METH_VARARGS | METH_KEYWORDS;
 
             capsule rec_capsule(rec, [](void *ptr) {

--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -306,7 +306,7 @@ protected:
             rec->def = new PyMethodDef();
             std::memset(rec->def, 0, sizeof(PyMethodDef));
             rec->def->ml_name = rec->name;
-            rec->def->ml_meth = reinterpret_cast<PyCFunction>(reinterpret_cast<void*>(*dispatcher));
+            rec->def->ml_meth = reinterpret_cast<PyCFunction>(reinterpret_cast<void (*) (void)>(*dispatcher));
             rec->def->ml_flags = METH_VARARGS | METH_KEYWORDS;
 
             capsule rec_capsule(rec, [](void *ptr) {

